src/server/active_tran_server.cpp

```diff
@@ -164,8 +162,38 @@ active_tran_server::push_request (ats_to_ps_request reqid, std::string &&payload
   m_ps_request_queue->push (reqid, std::move (payload));
 }
 
-void
-active_tran_server::receive_saved_lsa (cubpacking::unpacker &upk)
+void active_tran_server::receive_log_page (cubpacking::unpacker &upk)
+{
+  std::string message;
+  upk.unpack_string (message);
+
+  int error_code;
+  std::memcpy (&error_code, message.c_str (), sizeof (error_code));
+
+  LOG_PAGEID pageid = NULL_PAGEID;
+  std::string log_message = "Received log page message from Page Server. ";
+  if (error_code == NO_ERROR)
+    {
+      char log_page_buffer[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
+      LOG_PAGE *log_page = reinterpret_cast<LOG_PAGE *> (PTR_ALIGN (log_page_buffer, MAX_ALIGNMENT));
+
+      std::memcpy (log_page, message.c_str () + sizeof (error_code), db_log_page_size ());
+
+      pageid = log_page->hdr.logical_pageid;
+      log_message += ("Page ID: " + std::to_string (pageid) + '\n');
+    }
+  else
+    {
+      log_message += ("Error code: " + std::to_string (error_code) + '\n');
+    }
+
+  if (prm_get_bool_value (PRM_ID_ER_LOG_READ_LOG_PAGE))
+    {
+      _er_log_debug (ARG_FILE_LINE, log_message.c_str ());
+    }

```

Comment:
Rewrite suggestion, including next changes:
- The usual way to construct debug messages is to use the [fprintf format](https://en.cppreference.com/w/c/io/fprintf). Otherwise, if PRM_ID_ER_LOG_READ_LOG_PAGE is false, the message string is constructed for nothing. Remove log_message.
-  Remove pageid. It is used only once and log_page->hdr.logical_pageid can be used directly.
```suggestion
  if (error_code == NO_ERROR)
    {
      char log_page_buffer[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
      LOG_PAGE *log_page = reinterpret_cast<LOG_PAGE *> (PTR_ALIGN (log_page_buffer, MAX_ALIGNMENT));

      std::memcpy (log_page, message.c_str () + sizeof (error_code), db_log_page_size ());
	  if (prm_get_bool_value (PRM_ID_ER_LOG_READ_LOG_PAGE))
        {
		  _er_log_debug (ARG_FILE_LINE, "Received log page message from Page Server. Page ID: %lld\n",
			             log_page->hdr.logical_pageid);
        }
    }
  else
    {
	  if (prm_get_bool_value (PRM_ID_ER_LOG_READ_LOG_PAGE))
        {
		  _er_log_debug (ARG_FILE_LINE, "Received log page message from Page Server. Error code: %d\n", error_code);
        }
    }
```